File: snapo-desktop-electron/electron/network-event-filter.ts
import type { CdpMessage } from "../src/network/bridge-types.js";
import { matchesUrlFilterText } from "../src/network/url-filter.js";

export class NetworkEventFilter {
  private readonly matchingRequestIds = new Set<string>();

  constructor(private readonly searchText: string) {}

  matches(message: CdpMessage): boolean {
    if (this.searchText.trim().length === 0) return true;

    const requestId = stringAt(message.params, "requestId");
    const url = eventUrl(message);
    if (url != null && matchesUrlFilterText(url, this.searchText)) {
      if (requestId != null) this.matchingRequestIds.add(requestId);
      return true;
    }
    return requestId != null && this.matchingRequestIds.has(requestId);
  }
}

function eventUrl(message: CdpMessage): string | null {
  switch (message.method) {
    case "Network.requestWillBeSent":
      return stringAt(message.params, "request.url");
    case "Network.responseReceived":
      return stringAt(message.params, "response.url");
    case "Network.webSocketCreated":
      return stringAt(message.params, "url");
    default:
      return null;
  }
}

function stringAt(root: Record<string, unknown> | undefined, path: string): string | null {
  let current: unknown = root;
  for (const segment of path.split(".")) {
    if (!isRecord(current)) return null;
    current = current[segment];
  }
  return typeof current === "string" ? current : null;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value != null && !Array.isArray(value);
}

File: snapo-app-mac/SnapOCLI/NetworkEventFilter.swift
import Foundation
import SnapODeviceClient

struct NetworkEventFilter {
  private let tokens: URLFilterTokens
  private var matchingRequestIDs = Set<String>()

  init(_ searchText: String) {
    tokens = NetworkURLFilter.parseTokens(searchText)
  }

  mutating func matches(_ message: NetworkCDPMessage) -> Bool {
    if tokens.includes.isEmpty, tokens.excludes.isEmpty { return true }

    let requestID = string(at: "requestId", in: message.params)
    if let url = eventURL(message), NetworkURLFilter.matches(url: url, tokens: tokens) {
      if let requestID { matchingRequestIDs.insert(requestID) }
      return true
    }
    return requestID.map(matchingRequestIDs.contains) ?? false
  }

  private func eventURL(_ message: NetworkCDPMessage) -> String? {
    switch message.method {
    case "Network.requestWillBeSent":
      string(at: "request.url", in: message.params)
    case "Network.responseReceived":
      string(at: "response.url", in: message.params)
    case "Network.webSocketCreated":
      string(at: "url", in: message.params)
    default:
      nil
    }
  }
}
